File: app/uploader.go
package main

import (
    "fmt"
    "log"
    "os/exec"
    "os"
    "path/filepath"
    "net/http"
    "net/url"
    "strings"
    "encoding/json"
    "app/share"
)

const downloadFolder = "./downloads"
const queue_source = "upload"
const queue_destination = "image_download"

// VK API configuration
var (
    workDir              string
    vkAccessToken        string
    vkGroupID            string
    vkOwnerID            string
    videoSaveURL  = "https://api.vk.com/method/video.save?v=5.199"
)


func init() {
    vkAccessToken = os.Getenv("VK_ACCESS_TOKEN")
    vkGroupID = os.Getenv("VK_GROUP_ID")
    vkOwnerID = os.Getenv("VK_OWNER_ID")

    // Create download directory if it doesn't exist
    if _, err := os.Stat(downloadFolder); os.IsNotExist(err) {
        os.Mkdir(downloadFolder, os.ModePerm)
    }
    
    //Объявление используемых очередей
    share.DeclarateQueue(queue_source)
    share.DeclarateQueue(queue_destination)
}

func main() {
    share.RabbitConnect(queue_source, queue_destination, processMessage)
    share.CloseConnection()
}

func processMessage(video share.Video) (share.Video, error) {
    filePath, err := getLocalPathByVideoId(video.VideoID)
    if err != nil {
        log.Println(err)
        return video, err
    }
    log.Println("upload video: ", video.VideoID)

    data := url.Values{}
    data.Set("name", video.Title)
    data.Set("description", video.Description)
    data.Set("group_id", vkGroupID)

    r, err := http.NewRequest("POST", videoSaveURL, strings.NewReader(data.Encode()))
    r.Header.Add("Authorization", "Bearer " + vkAccessToken)
    r.Header.Add("Content-Type", "application/x-www-form-urlencoded")

    client := &http.Client{}
    urlResp, err := client.Do(r)
    if err != nil {
        return video, err
    }
    defer urlResp.Body.Close()

    var saveResponseR struct {
        Response struct {
            UploadURL string `json:"upload_url"`
            VideoID int `json:"video_id"`
        } `json:"response"`
    }
    err = json.NewDecoder(urlResp.Body).Decode(&saveResponseR)
    if err != nil {
        log.Println(err)
        return video, err
    }
    uploadURL := saveResponseR.Response.UploadURL
    vkVideoID := saveResponseR.Response.VideoID

    //ЗАПРОС 2 оправка файла на сервера VK

    cmd := exec.Command("curl", "-F", "'video_file=@"+filePath, uploadURL)
    err = cmd.Run()
    if err != nil {
        log.Println(err)
        return video, err
    }

    video.VkVideoID = vkVideoID

    return video, nil
}

func getLocalPathByVideoId(videoID string) (string, error) {
    fmt.Errorf(downloadFolder)
    files, err := filepath.Glob(filepath.Join(downloadFolder, fmt.Sprintf("%s.*", videoID)))
    if err != nil || len(files) == 0 {
        return "", fmt.Errorf("Empty file %s", videoID)
    }
    return files[0], nil
}

File: src/uploader.php
<?php


use GuzzleHttp\Client;
use PhpAmqpLib\Connection\AMQPStreamConnection;
use PhpAmqpLib\Message\AMQPMessage;

require(dirname(__FILE__) . '/../vendor/autoload.php');

$sourceQueue = 'upload';
$destinationQueue = 'wall_post';
$dir_save = '/app/data/downloads/';

$connection = new AMQPStreamConnection(
    $_ENV['RABBITMQ_HOST'],
    $_ENV['RABBITMQ_PORT'],
    $_ENV['RABBITMQ_DEFAULT_USER'],
    $_ENV['RABBITMQ_DEFAULT_PASS'],
    $_ENV['RABBITMQ_DEFAULT_VHOST'],
);

$channel = $connection->channel();
$channel->queue_declare($sourceQueue, false, true, false, false);
$channel->queue_declare($destinationQueue, false, true, false, false);
$channel->basic_qos(null, 1, null);
$channel->basic_consume($sourceQueue, '', false, false, false, false,
    function ($msg) use ($destinationQueue, $dir_save) {
        $body = $msg->body;
        $body = json_decode($body, true);

        $client = new Client([
            'base_uri' => 'https://api.vk.com',
            'headers' => [
                'Authorization' => 'Bearer ' . $_ENV['VK_ACCESS_TOKEN'],
                'Content-Type' => 'application/x-www-form-urlencoded',
            ]
        ]);
        $response = $client->post('/method/video.save', [
            'form_params' => [
                'v' => '5.119',
                'name' => $body['title'],
                'description' => $body['description'],
                'group_id' => $_ENV['VK_GROUP_ID'],
            ],
        ]);

        if ($response->getStatusCode() !== 200) {
            sleep(3600);
            return;
            //todo отправить сообщение об ошибке
        }

        $content = $response->getBody()->getContents();
        $content = json_decode($content, true);

        $vk_video_id = $content['response']['video_id'];
        $upload_url = $content['response']['upload_url'];

        $path_file = null;
        $file_name = null;
        $formats = ['mp4', 'mov', 'avi', 'wmv', 'flv', '3gp', 'webm', 'mkv'];

        foreach ($formats as $format) {
            $file_name = $body['video_id'] . '.' . $format;
            $path = $dir_save . $file_name;
            $exist = file_exists($path);
            if ($exist) {
                $path_file = $path;
                break;
            }
        }

        if (is_null($path_file)) {
            sleep(3600);
            return;
            //todo отправить сообщение об ошибке
        }

        $lastPercent = 0.0;
        $client2 = new Client();
        $response = $client2->post($upload_url, [
            'multipart' => [
                [
                    'name' => 'file',
                    'contents' => fopen($path_file, 'r'),
                    'filename' => $file_name
                ],
                [
                    'name' => 'key',
                    'contents' => 'value',
                ],
            ],
            'progress' => function ($downloadTotal, $downloaded, $uploadTotal, $uploaded) use (&$lastPercent) {
                if ($uploadTotal > 0) {
                    $percent = round(($uploaded / $uploadTotal) * 100, 1);
                    if ($percent !== $lastPercent) {
                        echo "Uploaded: $percent%\r";
                    }
                }
            },
        ]);

        if (!in_array($response->getStatusCode(), [200, 201])) {
            sleep(3600);
            //todo отправить сообщение об ошибке
        }

        $body['vk_video_id'] = $vk_video_id;

        $msg->delivery_info['channel']->basic_ack($msg->delivery_info['delivery_tag']);

        $msg_resp = new AMQPMessage(json_encode($body, JSON_UNESCAPED_UNICODE), ['delivery_mode' => AMQPMessage::DELIVERY_MODE_PERSISTENT]);

        //Отправляем сообщение
        $msg->delivery_info['channel']->basic_publish($msg_resp, '', $destinationQueue);

    });
while (count($channel->callbacks)) {
    $channel->wait();
}

